packages/ckeditor5-ckbox/src/ckboximageedit/ckboximageeditcommand.ts

```diff
@@ -0,0 +1,337 @@
+		const ckboxEditing = this.editor.plugins.get( CKBoxEditing );
+		const response: CKBoxRawAssetDataDefinition = await sendHttpRequest( {
+			url,
+			signal,
+			authorization: ckboxEditing.getToken().value
+		} );
+		const status = response.metadata!.metadataProcessingStatus;
+
+		if ( !status || status == 'queued' ) {
+			throw new Error( 'Image has not been processed yet.' );
+		}
+
+		return { data: { ...response } };
+	}
+
+	private async _waitForAssetProcessed( id: string, signal: AbortSignal ): Promise<CKBoxRawAssetDefinition> {
+		const result = await retry(
+			() => this._getAssetStatusFromServer( id, signal ),
+			{
+				signal,
+				maxAttempts: 5
+			}
+		);
+
+		if ( result.data.metadata!.metadataProcessingStatus != 'success' ) {
+			throw new Error( 'Image processing failed.' );
+		}
+
+		return result;
+	}
+
+	/**
+	 * Shows processing indicator while image is processing.
+	 *
+	 * @param asset Data about certain asset.
+	 */
+	private _showImageProcessingIndicator( asset: CKBoxRawAssetDefinition ): void {
+		const editor = this.editor;
+		const selectedImageElement = editor.model.document.selection.getSelectedElement()!;
+
+		editor.editing.view.change( writer => {
+			const imageElementView = editor.editing.mapper.toViewElement( selectedImageElement )!;
+			const imageUtils: ImageUtils = this.editor.plugins.get( 'ImageUtils' );
+			const img = imageUtils.findViewImgElement( imageElementView )!;
+
+			writer.removeStyle( 'aspect-ratio', img );
+			writer.setAttribute( 'width', asset.data.metadata!.width, img );
+			writer.setAttribute( 'height', asset.data.metadata!.height, img );
+
+			writer.setStyle( 'width', `${ asset.data.metadata!.width }px`, img );
+			writer.setStyle( 'height', `${ asset.data.metadata!.height }px`, img );
+
+			writer.addClass( 'image-processing', imageElementView );
+		} );
+	}
+
+	/**
+	 * Replace the edited image with the new one.
+	 */
+	private _replaceImage( element: ModelElement, asset: CKBoxRawAssetDefinition ) {
+		const editor = this.editor;
+		const imageCommand: InsertImageCommand = editor.commands.get( 'insertImage' )!;
+
+		const {
+			imageFallbackUrl,
+			imageSources,
+			imageWidth,
+			imageHeight,
+			imagePlaceholder
+		} = prepareImageAssetAttributes( asset );
+
+		const previousSelectionRanges = Array.from( editor.model.document.selection.getRanges() );
+
+		editor.model.change( writer => {
+			writer.setSelection( element, 'on' );
+
+			imageCommand.execute( {
+				source: {
+					src: imageFallbackUrl,
+					sources: imageSources,
+					alt: element.getAttribute( 'alt' ),
+					width: imageWidth,
+					height: imageHeight,
+					...( imagePlaceholder ? { placeholder: imagePlaceholder } : null )
+				}
+			} );
+
+			const previousChildren = element.getChildren();
+
+			element = editor.model.document.selection.getSelectedElement()!;
+
+			for ( const child of previousChildren ) {
+				writer.append( child, element );
```

Comment:
This is modifying the `$graveyard`. You must not modify the content of the graveyard. After undo the original caption could be lost or lead to an unexpected OT error. Maybe we could use `writer.cloneElement( child )`.